File: src/jsMain/kotlin/Backend.kt
import io.ktor.client.request.*
import io.ktor.client.statement.*
import io.ktor.http.*
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.encodeToJsonElement

object Backend {

  suspend fun tags(): Array<TagOverview> {
    return Json.decodeFromString(client.get("/api/tags").bodyAsText())
  }

  suspend fun snippets(tag: String): Array<SnippetOverview> {
    return Json.decodeFromString(client.get("/api/tag/$tag").bodyAsText())
  }

  suspend fun snippet(id: String): Snippet {
    return Json.decodeFromString(client.get("/api/snippet/$id").bodyAsText())
  }

  suspend fun deleteSnippet(id: String): Boolean {
    val response = client.delete("/api/snippet/${id}")
    Toast(response.bodyAsText())
    return response.status == HttpStatusCode.OK
  }

  suspend fun updateSnippet(snippet: Snippet): Snippet? {
    val response = client.post("/api/snippet/${snippet.id}") {
      contentType(ContentType.Application.Json)
      setBody(Json.encodeToJsonElement(snippet))
    }
    return if (response.status == HttpStatusCode.OK) {
      Toast("Snippet updated")
      Json.decodeFromString(response.bodyAsText())
    } else {
      Toast(response.bodyAsText())
      null
    }
  }

  suspend fun newSnippet(snippet: Snippet): Snippet? {
    val response = client.post("/api/new") {
      contentType(ContentType.Application.Json)
      setBody(Json.encodeToJsonElement(snippet))
    }
    return if (response.status == HttpStatusCode.OK) {
      Toast("Snippet created")
      updateSidebar(false)
      Json.decodeFromString(response.bodyAsText())
    } else {
      Toast(response.bodyAsText())
      null
    }
  }
}

File: static/js/backend.js
import * as Toast from '/assets/js/toast.js'
import * as Misc from '/assets/js/misc.js'

export async function tags() {
    return await fetch('/api/tags').then(res => res.json())
}

export async function byTag(tag) {
    return await fetch(`/api/tag/${tag}`).then(res => res.json())
}

export async function snippet(id) {
    return await fetch(`/api/snippet/${id}`).then(res => res.json())
}

export async function deleteSnippet(id) {
    return await fetch(`/api/snippet/${id}`, {
        method: 'DELETE'
    }).then(res => res.ok)
}

export async function updateSnippet(snippet) {
    let res = await fetch(`/api/snippet/${snippet.id}`, {
        method: 'POST',
        body: JSON.stringify(snippet)
    })
    if (res.status === 200) {
        Toast.show('Snippet updated')
        return await res.json()
    } else {
        Toast.show(await res.text())
    }
}

export async function newSnippet(snippet) {
    let res = await fetch('/api/new', {
        method: 'POST',
        body: JSON.stringify(snippet)
    })
    if (res.status === 200) {
        Toast.show('Snippet created')
        Misc.updateSidebar(false)
        return await res.json()
    } else {
        Toast.show(await res.text())
    }
}

export async function search(query) {
    return await fetch(`/api/search?q=${query}`).then(res => res.json())
}

export async function isAuthorized() {
    return await fetch('/authorized').then(res => res.text()).then(text => text === 'true')
}